File: iOS/TUIKit/TUICallKit/TUICallKit-Swift/ViewModel/Root/SingleCallViewModel.swift
//
//  SingleCallViewModel.swift
//
//
//  Created by vincepzhang on 2022/12/30.
//

import TUICallEngine

class SingleCallViewModel {
    
    let callStatusObserver = Observer()
    let mediaTypeObserver = Observer()
    let selfCallRoleObserver = Observer()
    let isShowFullScreenObserver = Observer()
    
    let selfCallStatus: Observable<TUICallStatus> = Observable(.none)
    let selfCallRole: Observable<TUICallRole> = Observable(.none)
    let mediaType: Observable<TUICallMediaType> = Observable(.unknown)
    let isShowFullScreen: Observable<Bool> = Observable(false)
    
    let enableFloatWindow = TUICallState.instance.enableFloatWindow
    
    init() {
        selfCallStatus.value = TUICallState.instance.selfUser.value.callStatus.value
        selfCallRole.value = TUICallState.instance.selfUser.value.callRole.value
        mediaType.value = TUICallState.instance.mediaType.value
        isShowFullScreen.value = TUICallState.instance.isShowFullScreen.value
        
        registerObserve()
    }
    
    deinit {
        TUICallState.instance.selfUser.value.callRole.removeObserver(selfCallRoleObserver)
        TUICallState.instance.selfUser.value.callStatus.removeObserver(callStatusObserver)
        TUICallState.instance.mediaType.removeObserver(mediaTypeObserver)
        TUICallState.instance.isShowFullScreen.removeObserver(isShowFullScreenObserver)
    }
    
    func registerObserve() {
        
        TUICallState.instance.selfUser.value.callRole.addObserver(selfCallRoleObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.selfCallRole.value = newValue
        })
        
        TUICallState.instance.selfUser.value.callStatus.addObserver(callStatusObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.selfCallStatus.value = newValue
        })
        
        TUICallState.instance.mediaType.addObserver(mediaTypeObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.mediaType.value = newValue
        })
        
        TUICallState.instance.isShowFullScreen.addObserver(isShowFullScreenObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.isShowFullScreen.value = newValue
        })
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/viewmodel/root/SingleCallViewModel.kt
package com.tencent.qcloud.tuikit.tuicallkit.viewmodel.root

import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.LiveData
import com.tencent.qcloud.tuikit.tuicallkit.data.User
import com.tencent.qcloud.tuikit.tuicallkit.state.TUICallState

class SingleCallViewModel {
    var callStatus = LiveData<TUICallDefine.Status>()
    var callRole = LiveData<TUICallDefine.Role>()
    var mediaType = LiveData<TUICallDefine.MediaType>()
    var isShowFullScreen = LiveData<Boolean>()
    var selfUser = LiveData<User>()

    init {
        callStatus = TUICallState.instance.selfUser.get().callStatus
        callRole = TUICallState.instance.selfUser.get().callRole
        mediaType = TUICallState.instance.mediaType
        isShowFullScreen = TUICallState.instance.isShowFullScreen
        selfUser = TUICallState.instance.selfUser
    }
}